### PR TITLE
Add latency metrics to mode2

### PR DESCRIPTION
**Pull request by leonorfmartins**
**What is this feature?**

Add latency metrics in mode2: in this mode, an error is returned if either of the stores fail.
To be merged after https://github.com/grafana/grafana/pull/87739

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Relates to https://github.com/grafana/search-and-storage-team/issues/29

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/apiserver/rest/dualwriter_mode1.go [hunk 1]
```diff
@@ -27,36 +33,191 @@ func NewDualWriterMode1(legacy LegacyStorage, storage Storage) *DualWriterMode1
...
+	ctx = klog.NewContext(ctx, log)
+	var method = "update"
+
+	startLegacy := time.Now().UTC()
+	res, async, errLegacy := d.Legacy.Update(ctx, name, objInfo, createValidation, updateValidation, forceAllowCreate, options)
+	if errLegacy != nil {
+		log.Error(errLegacy, "unable to update in legacy storage")
+		d.recordLegacyDuration(true, mode1, name, method, startLegacy)
+	}
+	d.recordLegacyDuration(false, mode1, name, method, startLegacy)
```
Comment (DanCech) [inline pkg/apiserver/rest/dualwriter_mode1.go:178]: `name` here is going to be the complete k8s name, we just want to pass the Kind (and only call recordLegacyDuration once)
```suggestion
		d.recordLegacyDuration(true, mode1, options.Kind, method, startLegacy)
	} else {
		d.recordLegacyDuration(false, mode1, options.Kind, method, startLegacy)
	}
```

**Comment by diegommm** [comment]
I would add the same observations about `context.Context` and `(time.Time).UTC()` usage as in https://github.com/grafana/grafana/pull/87739

File: pkg/apiserver/rest/dualwriter_mode1.go (hunk 1 above)
Comment (leonorfmartins) [reply pkg/apiserver/rest/dualwriter_mode1.go:178]: oh, I wasn't aware this would be the full k8s object! Should I also change in the other modes?

Comment (DanCech) [reply pkg/apiserver/rest/dualwriter_mode1.go:178]: Yep, we'll need to update that everywhere

File: pkg/apiserver/rest/dualwriter_mode2.go [hunk 2]
```diff
@@ -159,32 +189,40 @@ func (d *DualWriterMode2) DeleteCollection(ctx context.Context, deleteValidation
...
-		if !apierrors.IsNotFound(errUS) {
-			log.WithValues("objectList", deletedS).Error(errUS, "could not delete from duplicate storage")
+	startStorage := time.Now()
+	deletedS, _, err := d.Storage.Delete(ctx, name, deleteValidation, options)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.WithValues("objectList", deletedS).Error(err, "could not delete from duplicate storage")
 		}
 	}
+	d.recordStorageDuration(err != nil, mode2Str, options.Kind, method, startStorage)
```
Comment (DanCech) [inline pkg/apiserver/rest/dualwriter_mode2.go:225]: Right now we use this pattern in some places, and 2 separate calls in the cases where we `return` within the if block.  Would it make sense to move this before the if and use the same syntax everywhere?

**Review by DanCech — APPROVED** [review]

File: pkg/apiserver/rest/dualwriter_mode1.go [hunk 3]
```diff
@@ -129,7 +127,7 @@ func (d *DualWriterMode1) Delete(ctx context.Context, name string, deleteValidat
 		d.recordLegacyDuration(true, mode1Str, options.Kind, method, startLegacy)
 		return res, async, err
 	}
-	d.recordLegacyDuration(false, mode1Str, options.Kind, method, startLegacy)
+	d.recordLegacyDuration(false, mode1Str, name, method, startLegacy)
```
Comment (DanCech) [inline pkg/apiserver/rest/dualwriter_mode1.go:130]:
```suggestion
	d.recordLegacyDuration(false, mode1Str, options.Kind, method, startLegacy)
```

File: pkg/apiserver/rest/dualwriter_mode2.go [hunk 4]
```diff
@@ -34,12 +37,16 @@ func NewDualWriterMode2(legacy LegacyStorage, storage Storage) *DualWriterMode2
...
+	var method = "create"
 
+	startLegacy := time.Now()
 	created, err := d.Legacy.Create(ctx, obj, createValidation, options)
 	if err != nil {
 		log.Error(err, "unable to create object in legacy storage")
+		d.recordLegacyDuration(true, mode2Str, options.Kind, method, startLegacy)
 		return created, err
 	}
+	d.recordLegacyDuration(false, mode2Str, options.Kind, method, startLegacy)
```
Comment (DanCech) [inline pkg/apiserver/rest/dualwriter_mode2.go:49]:
```suggestion
	d.recordLegacyDuration(err != nil, mode2Str, options.Kind, method, startLegacy)
	if err != nil {
		log.Error(err, "unable to create object in legacy storage")
		return created, err
	}
```

**Pull request merged**